Update setup for developer-centric scripts

razvand opened the pull request:
Update setup to include scripts for building and running. Restructure contents to use the `workdir/` directory as the directory to store repository clones and build output; use `defconfigs/` to store default configuration files.

razvand committed:
Use `workdir/` as clone directory instead of `.unikraft/`

`.unikraft/` is used as the directory to clone library repositories
(including the `unikraft` core repository). Being a hidden directory, it
is not immediately available to the developer.

Use `workdir/` instead. Also use `workdir/build/` as output directory
and ignore the `wokdir/` directory.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
kraft.yaml: Add kraftkit target names

Add explicit target names for the KraftKit tagets in `kraft.yaml`. Add
all combinations of platforms (QEMU, Firecracker), filesystems (9pfs,
initrd) and architectures (x86_64, arm64).

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Move and config files to defconfigs/ directory

Move all the defconfig files into the `defconfigs/` directory.
Also minimise them using `make savedefconfig`, removing all
unnecesary config options. Add all deconfig files for all combinations
of filesystems, architectures and platforms.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Introduce support for generated scripts

Add `scripts/` directory that provides support for generating build and
run scripts. Generator script is stored externally, in the `app-testing`
repository.

The `run.yaml` file defines the possible running configurations.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
.gitignore: Ignore all output entries

Update `.gitignore` to ignore all output entries: configuration files,
`rootfs/` directory, CPIO temporary file.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
README.md: Update instructions

Add instructions for building using KraftKit. Add instruction on using
build and run scripts. Update commands and paths to feature the new
setup: `workdir/` for clones and build output, `defconfigs/` for default
configurations.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

StefanJum commented on README.md:
```diff
@@ -13,37 +53,30 @@ For building and running everything for `x86_64`, follow the steps below:
-mkdir fs0/
-tar -C fs0/ -xvf rootfs.tar.gz
-mkdir .unikraft
-git clone https://github.com/unikraft/unikraft .unikraft/unikraft
-git clone https://github.com/unikraft/lib-python3 .unikraft/libs/python3
-git clone https://github.com/unikraft/lib-musl .unikraft/libs/musl
-git clone https://github.com/unikraft/lib-lwip .unikraft/libs/lwip
-git clone https://github.com/unikraft/lib-compiler-rt .unikraft/libs/compiler-rt
```
Leave the setup commands in.

StefanJum commented on README.md:
```diff
@@ -13,37 +53,30 @@ For building and running everything for `x86_64`, follow the steps below:
+git clone https://github.com/unikraft/app-python3 python3
+cd python3/
+wget https://raw.githubusercontent.com/unikraft/app-testing/staging/scripts/generate.py -O scripts/generate.py
+chmod a+x scripts/generate.py
```
Add the cloning commands here.

StefanJum commented on README.md:
````diff
@@ -109,77 +150,36 @@ Follow the steps below for the setup:
 
      ```console
+     git clone https://github.com/unikraft/unikraft workdir/unikraft
      git clone https://github.com/unikraft/lib-python3 libs/python3
````
```suggestion
     git clone https://github.com/unikraft/lib-python3 workdir/libs/python3
```

StefanJum commented on README.md:
```diff
@@ -109,77 +150,36 @@ Follow the steps below for the setup:
+     git clone https://github.com/unikraft/unikraft workdir/unikraft
      git clone https://github.com/unikraft/lib-python3 libs/python3
-
      git clone https://github.com/unikraft/lib-musl libs/musl
```
```suggestion
     git clone https://github.com/unikraft/lib-musl workdir/libs/musl
```

StefanJum commented on README.md:
```diff
@@ -109,77 +150,36 @@ Follow the steps below for the setup:
-
      git clone https://github.com/unikraft/lib-musl libs/musl
-
      git clone https://github.com/unikraft/lib-lwip libs/lwip
```
```suggestion
     git clone https://github.com/unikraft/lib-lwip workdir/libs/lwip
```

StefanJum commented on README.md:
```diff
@@ -109,77 +150,36 @@ Follow the steps below for the setup:
-
      git clone https://github.com/unikraft/lib-lwip libs/lwip
-
      git clone https://github.com/unikraft/lib-compiler-rt libs/compiler-rt
```
```suggestion
     git clone https://github.com/unikraft/lib-compiler-rt workdir/libs/compiler-rt
```

StefanJum reviewed (changes requested):
See the few comments.

razvand committed:
Introduce setup script to clone repositories

Introduce `scripts/setup.sh` script to clone repositories in the
`workdir/` directory.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

StefanJum reviewed (approved):
All good.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>
